avocado/core/varianter.py

```diff
@@ -386,51 +326,45 @@ def __init__(self, debug=False):
                in order to provide the right results.
         """
         self.default_params = {}
-        self.variants = None
         self.debug = debug
-        self.data = tree.TreeNodeDebug() if debug else tree.TreeNode()
-        self._mux_path = None
+        self.node_class = tree.TreeNode if not debug else tree.TreeNodeDebug
         self.ignore_new_data = False    # Used to ignore new data when parsed
+        self._variant_plugins = dispatcher.VarianterDispatcher()
+        self._no_variants = None
 
     def parse(self, args):
         """
         Apply options defined on the cmdline
 
         :param args: Parsed cmdline arguments
         """
-        root = copy.deepcopy(self._process_default_params(args))
-        if self.data != tree.TreeNode():
-            # TODO: Do this per-variant-plugin
-            root.merge(self.data)
-            self.variants = MuxTree(root)
-        else:
-            self.variants = False
-        self._mux_path = getattr(args, 'mux_path', None)
-        if self._mux_path is None:
-            self._mux_path = ['/run/*']
+        defaults = self._process_default_params(args)
+        self._variant_plugins.map_method_copy("initialize", args)
+        self._variant_plugins.map_method_copy("update_defaults", defaults)
+        self._no_variants = sum(self._variant_plugins.map_method("__len__"))
 
     def _process_default_params(self, args):
         """
         Process the default params
 
         :param args: Parsed cmdline arguments
         """
-        default_params = tree.TreeNode()
+        default_params = self.node_class()
         for default_param in self.default_params.itervalues():
             default_params.merge(default_param)
         self.default_params = default_params
         # FIXME: Backward compatibility params, to be removed when 36 LTS is
         # discontinued
         if (not getattr(args, "mux_skip_defaults", False) and
                 hasattr(args, "default_avocado_params")):
-            self.default_params.merge(args.default_avocado_params)
+            self.default_params.merge(args.default_avocado_params)  # self.default_params is TreeNode pylint: disable=E1101
```

Comment:
I don't remember you already mention that as you did not. You just committed several commits which were squashed to already quite long commits and referenced the tree in the PR. It is not really feasible to go and diff all diffs you committed somewhere. The fundament of github is that one can comment on lines of commits which is suppose to simplify or speed-up the review and I'm really not able to follow the other approach you choose. If you want some changes please mention it in-line so I can go through the suggestions, fix it and eventually respond that I fixed it (which helps to avoid issues like "I already mentioned it here...".

Anyway to respond to the question, it does not clashes. Python allows this construct and I used it here because I saw it as a best option. Another way would be to have `self.default_params` and `self._default_params` where the `default_params` would be publicly available variable to register your default params and the `_default_params` would be `None` (consuming size of an attribute and None). After the `process_default_params` the `default_params` would be erased (still consuming the size of an empty dict and attribute) and `_default_params` would be the produced `default_params`.

I choose to overwrite the same attribute to avoid the unnecessary attributes, but if you really dislike it I can split these into two variables.